File: mini-util/src/main/java/com/mini/core/util/holder/ClassHolder.java
package com.mini.core.util.holder;


import java.lang.annotation.Annotation;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public final class ClassHolder<T> {
    private static final Map<Class<?>, ClassHolder<?>> M = new HashMap<>();
    private final Map<String, FieldHolder<T>> fields = new HashMap<>();
    private final Class<T> type;
    private T instance;


    private ClassHolder(Class<T> type) {
        this.type = type;
    }

    public final String getName() {
        return type.getSimpleName();
    }

    public Class<T> getType() {
        return type;
    }

    public final T getInstance() throws ReflectiveOperationException {
        if (Objects.isNull(ClassHolder.this.instance)) {
            var constructor = type.getConstructor();
            instance = constructor.newInstance();
        }
        return instance;
    }

    public Map<String, FieldHolder<T>> getFields() {
        return fields;
    }

    public void addField(FieldHolder<T> field) {
        this.fields.put(field.getName(), field);
    }

    public <A extends Annotation> A[] getAnnotationsByType(Class<A> clazz) {
        return type.getAnnotationsByType(clazz);
    }

    public <A extends Annotation> A getAnnotation(Class<A> clazz) {
        return type.getAnnotation(clazz);
    }

    @SuppressWarnings("unchecked")
    public static <T> ClassHolder<T> create(Class<T> type) {
        return (ClassHolder<T>) M.computeIfAbsent(type, key -> {
            ClassHolder<T> holder = new ClassHolder<>(type);
            FieldHolder.create(holder);
            return holder;
        });
    }


}

File: mini-util/src/main/kotlin/com/mini/core/holder/ClassHolder.kt
@file:Suppress("unused")

package com.mini.core.holder

import kotlin.reflect.KClass

class ClassHolder<T : Any> private constructor(val type: KClass<in T>) {
    val fields: MutableMap<String, FieldHolder<T>> = mutableMapOf()

    @Suppress("UNCHECKED_CAST")
    val instance: T? by lazy {
        type.constructors.firstOrNull { it.parameters.isEmpty() }?.call() as? T
    }

    inline fun <reified A : Annotation> getAnnotationsByType(clazz: KClass<in A>): Array<A> {
        return type.annotations.filter {
            clazz.isInstance(it)
        }.map { it as A }.toTypedArray()
    }

    inline fun <reified A : Annotation> getAnnotation(clazz: KClass<in A>): A? {
        return type.annotations.find {
            clazz.isInstance(it)
        } as? A
    }

    companion object {
        private val M: MutableMap<KClass<*>, ClassHolder<*>> = mutableMapOf()

        @Synchronized
        @Suppress("UNCHECKED_CAST")
        fun <T : Any> create(type: KClass<T>): ClassHolder<T> {
            return M.computeIfAbsent(type) {
                val holder = ClassHolder(type)
                // holder.fields.put()

                return@computeIfAbsent holder
            } as ClassHolder<T>
        }
    }
}